File: python/composio/tools/env/e2b/workspace.py
"""E2B Workspace."""

import os
import time
import typing as t
from dataclasses import dataclass
from uuid import uuid4

from composio.exceptions import ComposioSDKError
from composio.tools.env.base import RemoteWorkspace, WorkspaceConfigType


try:
    import requests
    from e2b import Sandbox

    E2B_INSTALLED = True
except ImportError:
    Sandbox = t.Any
    E2B_INSTALLED = False


TOOLSERVER_PORT = 8000
TOOLSERVER_URL = "https://{host}/api"

DEFAULT_TEMPLATE = "2h9ws7lsk32jyow50lqz"
ENV_E2B_TEMPLATE = "E2B_TEMPLATE"


@dataclass
class Config(WorkspaceConfigType):
    """Host configuration type."""

    template: t.Optional[str] = None
    """Template ID for creating the sandbox, if not provided the composio tooling server template will be used."""

    api_key: t.Optional[str] = None
    """E2B API Key."""

    port: t.Optional[int] = None
    """Port for launching the toolserver on the E2B sandbox."""


class E2BWorkspace(RemoteWorkspace):
    """Create and manage E2B workspace."""

    sandbox: Sandbox

    def __init__(self, config: Config):
        """Initialize E2B workspace."""
        if not E2B_INSTALLED:
            raise ComposioSDKError(
                "`e2b` is required to use e2b workspace, "
                "run `pip3 install composio-core[e2b]` or "
                "`pip3 install e2b e2b-code-interpreter` to install",
            )

        super().__init__(config=config)
        template = config.template
        if template is None:
            template = os.environ.get(ENV_E2B_TEMPLATE)
            if template is not None:
                self.logger.debug(f"Using E2B template `{template}` from environment")
            template = template or DEFAULT_TEMPLATE

        self.template = template
        self.api_key = config.api_key
        self.port = config.port or TOOLSERVER_PORT

    def _wait(self) -> None:
        deadline = time.time() + float(os.environ.get("WORKSPACE_WAIT_TIMEOUT", 60.0))
        while time.time() < deadline:
            try:
                if (
                    self._request(endpoint="", method="get", log=False).status_code
                    == 200
                ):
                    return
            except requests.ConnectionError:
                time.sleep(1)

    def setup(self) -> None:
        """Start toolserver."""
        # Start sandbox
        self.sandbox = Sandbox(template=self.template, api_key=self.api_key)
        self.url = TOOLSERVER_URL.format(host=self.sandbox.get_host(self.port))
        self.logger.debug(f"{self}.url = {self.url}")

        # Start app update in background
        self.sandbox.commands.run(
            cmd="composio apps update",
            envs=self.environment,
            background=False,
        )

        # Setup SSH server
        _ssh_password = uuid4().hex.replace("-", "")
        self.sandbox.commands.run(
            cmd=f"echo user:{_ssh_password} | sudo chpasswd",
            envs=self.environment,
            background=False,
        )
        self.sandbox.commands.run(
            cmd="sudo service ssh restart",
            envs=self.environment,
            background=False,
        )
        self.sandbox.commands.run(
            cmd=(
                f"COMPOSIO_LOGGING_LEVEL=debug "
                f"_SSH_USERNAME=user _SSH_PASSWORD={_ssh_password} "
                f"composio serve -h '0.0.0.0' -p {self.port}"
            ),
            envs=self.environment,
            background=True,
        )
        self.host = self.sandbox.get_host(port=80)
        self.ports = []
        self._wait()

    def teardown(self) -> None:
        """Teardown E2B workspace."""
        super().teardown()
        if hasattr(self, "sandbox"):
            self.sandbox.kill()

File: js/src/env/e2b/workspace.ts
import { RemoteWorkspace } from "../base";
import { getUUID } from "../../utils/getUUID";
import { getEnvVariable, nodeExternalRequire } from "../../utils/shared";
import { Sandbox } from "e2b";
import { IWorkspaceConfig, WorkspaceConfig } from "../config";
import logger from "../../utils/logger";

const DEFAULT_TEMPLATE = "2h9ws7lsk32jyow50lqz";
const TOOLSERVER_PORT = 8000;
const TOOLSERVER_URL = "https://{host}/api";

const ENV_E2B_TEMPLATE = "E2B_TEMPLATE";

export type IE2BConfig = IWorkspaceConfig & {
  template?: string;
  apiKey?: string;
  port?: number;
};

export class E2BWorkspace extends RemoteWorkspace {
  sandbox: Sandbox | undefined;
  template: string;
  apiKey?: string;
  port: number;

  constructor(configRepo: WorkspaceConfig<IE2BConfig>) {
    super(configRepo);
    this.template =
      configRepo.config.template ||
      getEnvVariable(ENV_E2B_TEMPLATE, DEFAULT_TEMPLATE)!;
    this.apiKey = configRepo.config.apiKey;
    this.port = configRepo.config.port || TOOLSERVER_PORT;
  }

  async setup(): Promise<void> {
    this.sandbox = await Sandbox.create({
      template: this.template,
      envVars: this.environment,
      apiKey: this.apiKey,
    });

    this.url = TOOLSERVER_URL.replace(
      "{host}",
      await this.sandbox!.getHostname(this.port)
    );

    logger.debug("E2B Composio Server is live at " + this.url);

    const process = await this.sandbox!.process.start({
      cmd: "composio apps update",
    });

    await process.wait();

    const _ssh_username = getUUID().replace(/-/g, "");
    const _ssh_password = getUUID().replace(/-/g, "");

    await this.sandbox!.process.start({
      cmd: `sudo useradd -rm -d /home/${_ssh_username} -s /bin/bash -g root -G sudo ${_ssh_username}`,
    });

    await this.sandbox!.process.start({
      cmd: `echo ${_ssh_username}:${_ssh_password} | sudo chpasswd`,
    });

    await this.sandbox!.process.start({
      cmd: "sudo service ssh restart",
    });

    await this.sandbox!.process.start({
      cmd: `_SSH_USERNAME=${_ssh_username} _SSH_PASSWORD=${_ssh_password} COMPOSIO_LOGGING_LEVEL=debug composio serve -h '0.0.0.0' -p ${this.port}`,
    });

    await this.waitForSpinUp();
  }

  async waitForSpinUp() {
    const timeout = 2 * 60 * 1000; // 2 minutes in milliseconds
    const startTime = Date.now();
    let status = 0;

    while (status !== 200) {
      if (Date.now() - startTime > timeout) {
        await this.teardown();
        throw new Error("Timeout: Server did not spin up within 2 minutes");
      }

      try {
        status = (await this._request("", "get")).status;
      } catch (error) {}

      if (status !== 200) {
        await new Promise((resolve) => setTimeout(resolve, 1000));
      }
    }
  }

  async teardown(): Promise<void> {
    if (!this.sandbox) {
      throw new Error("Sandbox not initialized");
    }
    await super.teardown();
    await this.sandbox.close();
  }
}
